### ISSUE TITLE
Piano roll toolbar is too wide

### ISSUE THREAD

**Issue by michaelgregorius** [LMMS/lmms#2286]
The current piano roll toolbar is very wide and create problems on size constraint displays, e.g. 1280x1024 or smaller:
![pianoroll-wide](https://cloud.githubusercontent.com/assets/9293269/9429178/f34d4ce8-49c7-11e5-993a-ea1482cef2c1.png)
The individual groups should be broken down into separate toolbars with a two line default layout.

**Comment by tresf** [comment]
https://github.com/LMMS/lmms/issues/387 :+1:

**Comment by michaelgregorius** [comment]
The patch above changes the piano roll's toolbar to the following look:
![pianorollchanges-movabletoolsbars](https://cloud.githubusercontent.com/assets/9293269/9428756/25249618-49b9-11e5-9f12-4c9638e28cf1.png)
There are now five groups / tool bars: "Transport controls", "Note controls",
"Copy paste controls", "Timeline controls" and "Zoom and note controls"
(this group is a bit mixed). Each group can be turned off and on using
the standard toolbar context menu provided by Qt.

The new default layout of the toolbars saves horizontal space by putting
the "Zoom and note controls" below the other toolbars. The toolbars can
be moved but their positions are not saved. Floating toolbars are
disabled, i.e. the toolbars will always stay inside the window (top,
bottom, left, right).

**Comment by tresf** [comment]
Moving the conversation here... 

> @tresf This is all done using stuff that's already available from Qt! If you look at the code changes you will see that it was a rather low hanging fruit. There is some not so nice code concerning `m_toolBar` from the `Editor` class because I had to override some of the defaults being taken in the `Editor` implementation. I don't think it's a good idea that every `Editor` automatically has some transport controls but for now this does not seem to create any problems (except some headscratching when reading some of the code :wink:).

Wishful thinking... should we just fix the parent Editor once and for all?

-Tres

**Comment by michaelgregorius** [comment]
@tresf I had a closer look at the `Editor` class. It seems that it's only concerned with transport functionality and providing the basis for toolbars by inheriting from `QMainWindow`. I thought that it would provide a bit more functionality which might be useful to other editors that might not need transport capabilities. So if we'd remove the transport toolbar we could more or less remove the whole class.

But whatever the decision is I think all these things should be handled in a new issue anyway because it would not only concern the piano roll.

**Comment by Umcaruje** [comment]
Very cool looking! 

>The toolbars can be moved but their positions are not saved

That's a real bummer for me... Is there a way these positions could be saved in the .lmmsrc file? And the hidden/visible toolbars.

***

Also, here is an idea - would locking of the toolbar be possible? e.g. no moving of toolbars and the drag handles would be gone.

**Comment by midi-pascal** [comment]
Don't know if its related to this issue but recently (few days in master) the default size of the piano roll window changed in strange way.
To reproduce:
1- Create a new project
2- Add an instrument - Sf2 Player for the purpose -
3- Double click its track
The result looks really ugly - IMHO -:
![pianoroll](https://cloud.githubusercontent.com/assets/6962451/9429430/993cde8c-499d-11e5-8943-d3a61f96a008.jpg)
Should not the initial size of the Piano roll window minimally allow to see the menus and toolbars?

**Comment by michaelgregorius** [comment]
@midi-pascal Yes, I also noticed this. The automation editor shows the same problem by the way. I propose to create a new issue to cover this.

**Comment by tresf** [comment]
> So if we'd remove the transport toolbar we could more or less remove the whole class.

I see.  But according to the comments, this offers a common interface for both BBEditor and SongEditor as well, right?  Or are you saying architecturally they're just widgets and we don't benefit at all from having them inherit the same class?

-Tres

**Comment by michaelgregorius** [comment]
@Umcaruje Yes, I guess both proposals should be possible to implement but might be rather complex (especially the first one).

However, I think several small improvements are better than one large improvement. So I'd propose to create two new issues once the above pull request is merged (in case it will be accepted). Thanks!

**Comment by tresf** [comment]
Also, I like the idea of a barebones editor class, since most of our editors carry similar layout and transport controls, although if we just add transport as a toolbar item, it would probably be easier to wrap our head around where and how components are added in each editor class, right?

**Comment by tresf** [comment]
@Umcaruje I agree with @michaelgregorius, this is to fix the overflow/wrapping problem and the position and visibility can be offered as a separate enhancement.  I too find this fix to be valuable enough to take as-is as there are many small adjustments to the interface that don't persist across sessions, and living with them is easier than having a subwindow that is stuck at 900pixels, IMO. :)

**Comment by michaelgregorius** [comment]
@tresf I think my concerns about the `Editor` class can best be described by the design principle ["Favor composition over inheritance"](http://stackoverflow.com/questions/49002/prefer-composition-over-inheritance). Creating an explicit transport toolbar class and using it in the other editor classes would likely provide more flexibility with regards to the behavior. I also think that providing a `bool` to the `Editors` constructor is not very descriptive of what it's doing (hide record functionality). On the other hand if you could call something like `m_transportToolBar->enableRecording( false )` in the other editors' constructors it would be more obvious in which way the transport toolbar is configured and what's happening.

**Comment by tresf** [comment]
@michaelgregorius agreed wholeheartedly.  I read a lot of code written by a lot of people.  Inheritance is nice when you have the full-bells-and-whistles IDE at your fingertips, but no so much when you're in `vi`, `gedit`, `kate`, `nano`, `notepad`, etc.  In this case, I agree with both proposals, including that of the obscure constructor.

-Tres

**Comment by tresf** [comment]
So a bit more on topic, I'll rephrase my previous question... Ignoring whether or not from a design perspective the question of whether or not the  `Editor` class **should be the parent of all Editors**... 

Shall we just fix all of the toolbars now?

-Tres

**Comment by tresf** [comment]
**Edit:** If so, I'd also be happy marking #387 as a duplicate of this and filing a new request for a collapsable menu bar. :smiley:

**Comment by michaelgregorius** [comment]
@tresf Ah, ok, I somehow must have missed that question. :smiley: It seems that this would encompass fixing the "Automation Editor", the "Beats+Bassline Editor" and the "Song Editor". I can give it a shot in the next few days. This should hopefully be quite similar to the changes made for the piano roll.

**Comment by tresf** [comment]
> Shall I push successful changes onto #2287?

If the code is related enough to warrant combining them, yes.  The issue and PR titles can easily be changed.

**Comment by tresf** [comment]
> It seems that this would encompass fixing the "Automation Editor", the "Beats+Bassline Editor" and the "Song Editor"

Ok, so wishful thinking on my part that this could be extended to the main toolbar as well.  The main toolbar suffers this same problem IMO, but the real-estate problem is vertical rather than horizontal.  Perhaps that is best left as its own enhancement then.

**Comment by midi-pascal** [comment]
@michaelgregorius Ok. Opened #2288 then :smile:

**Comment by michaelgregorius** [comment]
@tresf I have switched the other editors to using movable tool bars as well (see #2287).

I also think that the changes to the main tool bar should be an issue of its own.

**Comment by tresf** [comment]
:+1:

**Comment by michaelgregorius** [comment]
Closed via merged #2287.
